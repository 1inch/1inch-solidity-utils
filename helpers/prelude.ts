import chai, {Assertion, AssertionError,assert,expect, config,should} from 'chai';
import 'chai-bn';
import chaiAsPromised from 'chai-as-promised';
import { toBN } from 'web3-utils';
chai.use(chaiAsPromised);
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { time, ether } = require('@openzeppelin/test-helpers');

export const constants = {
    ZERO_ADDRESS: '0x0000000000000000000000000000000000000000',
    ZERO_BYTES32: '0x0000000000000000000000000000000000000000000000000000000000000000',
    MAX_UINT256: toBN('2').pow(toBN('256')).sub(toBN('1')).toString(),
    MAX_INT256: toBN('2').pow(toBN('255')).sub(toBN('1')).toString(),
    MIN_INT256: toBN('2').pow(toBN('255')).mul(toBN('-1')).toString(),
} as const;

// utils
export {
    toBN
}

//test-helpers
export {
    time,
    ether
}

// chai
export {
    Assertion,
    AssertionError,
    assert,
    expect, 
    config,
    should
}