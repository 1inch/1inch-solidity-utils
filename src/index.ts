// created from 'create-ts-index'

export * from './expect';
export * from './permit';
export * from './prelude';
export * from './profileEVM';
export * from './utils';
