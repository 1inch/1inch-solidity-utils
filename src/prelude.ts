import chai, { Assertion, AssertionError, assert, expect, config, should } from 'chai';
import 'chai-bn';
import chaiAsPromised from 'chai-as-promised';
import { toWei } from 'web3-utils';
import BN from 'bn.js';
chai.use(chaiAsPromised);
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { time: timeImpl } = require('@openzeppelin/test-helpers');

export function toBN (num: string | number, base?: number | 'hex'): BN {
    if (typeof(num) === 'string' && num.startsWith('0x')) {
        return new BN(num.substring(2), 16);
    }
    return new BN(num, base);
}

export const constants = {
    ZERO_ADDRESS: '0x0000000000000000000000000000000000000000',
    EEE_ADDRESS: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE',
    ZERO_BYTES32: '0x0000000000000000000000000000000000000000000000000000000000000000',
    MAX_UINT256: toBN('2').pow(toBN('256')).sub(toBN('1')).toString(),
    MAX_INT256: toBN('2').pow(toBN('255')).sub(toBN('1')).toString(),
    MIN_INT256: toBN('2').pow(toBN('255')).mul(toBN('-1')).toString(),
} as const;

// utils
export {
    BN,
};

//test-helpers
export type Time = {
    increaseTo: (target: string | number | BN) => Promise<BN>,
    latest: () => Promise<BN>
}

export const time: Time = timeImpl;

export function ether (n: string): BN {
    return toBN(toWei(n, 'ether'));
}

// chai
export {
    Assertion,
    AssertionError,
    assert,
    expect,
    config,
    should,
};
