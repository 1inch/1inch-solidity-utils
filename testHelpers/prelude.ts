import chai, {Assertion, AssertionError,assert,expect, config,should} from 'chai';
import 'chai-bn';
import chaiAsPromised from 'chai-as-promised';
chai.use(chaiAsPromised);

export {
    Assertion,
    AssertionError,
    assert,
    expect, 
    config,
    should
}