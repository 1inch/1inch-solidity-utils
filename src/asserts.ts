import { expect, toBN } from './prelude';
import BN from "bn.js";

export function toBNExtended (value: string | number | BN): BN {
    if (typeof value === 'string' || typeof value === 'number') {
        return toBN(value);
    }
    return value;
}

export function assertRoughlyEqualValues (expected: string | number | BN, actual: string | number | BN, relativeDiff: number) {
    const expectedBN = toBNExtended(expected);
    const actualBN = toBNExtended(actual);

    let multiplerNumerator = relativeDiff;
    let multiplerDenominator = toBN('1');
    while (!Number.isInteger(multiplerNumerator)) {
        multiplerDenominator = multiplerDenominator.mul(toBN('10'));
        multiplerNumerator *= 10;
    }
    const diff = expectedBN.sub(actualBN).abs();
    const treshold = expectedBN.mul(toBN(multiplerNumerator.toString())).div(multiplerDenominator);
    if (!diff.lte(treshold)) {
        expect(actualBN).to.be.bignumber.equal(expectedBN, `${actualBN} != ${expectedBN} with ${relativeDiff} precision`);
    }
}
