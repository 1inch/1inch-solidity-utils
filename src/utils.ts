import { constants } from './prelude';
import { time } from '@nomicfoundation/hardhat-network-helpers';
import { ethers } from 'hardhat';
import { BaseContract, BigNumber, Bytes, CallOverrides, ContractTransaction } from 'ethers';
import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';

export async function timeIncreaseTo (seconds: number | string) {
    const delay = 1000 - new Date().getMilliseconds();
    await new Promise(resolve => setTimeout(resolve, delay));
    await time.increaseTo(seconds);
}

export interface Token extends BaseContract {
    balanceOf(
        account: string,
        overrides?: CallOverrides
    ): Promise<BigNumber>;
}

export async function trackReceivedTokenAndTx<T extends unknown[]> (
    token: Token | {address: typeof constants.ZERO_ADDRESS} | {address: typeof constants.EEE_ADDRESS},
    wallet: string,
    txPromise: (...args: T) => Promise<ContractTransaction>,
    ...args: T
) {
    const [balanceFunc, isETH] =
            'balanceOf' in token
                ? [() => token.balanceOf(wallet), false]
                : [async () => (await ethers.provider.getBalance(wallet)), true];
    const preBalance = await balanceFunc();
    const txResult = await txPromise(...args);
    const txReceipt = await txResult.wait();
    const txFees = (wallet.toLowerCase() === txResult.from.toLowerCase() && isETH)
        ? txReceipt.gasUsed.toBigInt() * txReceipt.effectiveGasPrice.toBigInt()
        : 0n;
    const postBalance = await balanceFunc();
    return [postBalance.sub(preBalance).add(txFees), txResult] as const;
}

export function fixSignature (signature: string) {
    // in geth its always 27/28, in ganache its 0/1. Change to 27/28 to prevent
    // signature malleability if version is 0/1
    // see https://github.com/ethereum/go-ethereum/blob/v1.8.23/internal/ethapi/api.go#L465
    let v = parseInt(signature.slice(130, 132), 16);
    if (v < 27) {
        v += 27;
    }
    const vHex = v.toString(16);
    return signature.slice(0, 130) + vHex;
}

export async function signMessage (signer: SignerWithAddress, messageHex: string | Bytes = '0x') {
    return fixSignature(await signer.signMessage(messageHex));
}

export async function countInstructions (txHash: string, instructions: string[]) {
    const trace = await ethers.provider.send('debug_traceTransaction', [ txHash ]);

    const str = JSON.stringify(trace);

    return instructions.map(instr => {
        return str.split('"' + instr.toUpperCase() + '"').length - 1;
    });
}
