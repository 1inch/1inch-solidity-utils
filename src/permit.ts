import { SignTypedDataVersion, TypedDataUtils } from '@metamask/eth-sig-util';
import { constants } from './prelude';
import { Contract } from 'ethers';
import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';
import { splitSignature } from 'ethers/lib/utils';

export const TypedDataVersion = SignTypedDataVersion.V4;
export const defaultDeadline = constants.MAX_UINT256;

export const EIP712Domain = [
    { name: 'name', type: 'string' },
    { name: 'version', type: 'string' },
    { name: 'chainId', type: 'uint256' },
    { name: 'verifyingContract', type: 'address' },
];

export const Permit = [
    { name: 'owner', type: 'address' },
    { name: 'spender', type: 'address' },
    { name: 'value', type: 'uint256' },
    { name: 'nonce', type: 'uint256' },
    { name: 'deadline', type: 'uint256' },
];

export const DaiLikePermit = [
    { name: 'holder', type: 'address' },
    { name: 'spender', type: 'address' },
    { name: 'nonce', type: 'uint256' },
    { name: 'expiry', type: 'uint256' },
    { name: 'allowed', type: 'bool' },
];

export function trim0x(bigNumber: bigint | string) {
    const s = bigNumber.toString();
    if (s.startsWith('0x')) {
        return s.substring(2);
    }
    return s;
}

export function cutSelector(data: string) {
    const hexPrefix = '0x';
    return hexPrefix + data.substr(hexPrefix.length + 8);
}

export function domainSeparator(name: string, version: string, chainId: string, verifyingContract: string) {
    return (
        '0x' +
        TypedDataUtils.hashStruct(
            'EIP712Domain',
            { name, version, chainId, verifyingContract },
            { EIP712Domain },
            TypedDataVersion,
        ).toString('hex')
    );
}

export function buildData(
    name: string,
    version: string,
    chainId: number,
    verifyingContract: string,
    owner: string,
    spender: string,
    value: string,
    nonce: string,
    deadline: string = defaultDeadline.toString(),
) {
    return {
        types: { Permit },
        domain: { name, version, chainId, verifyingContract },
        message: { owner, spender, value, nonce, deadline },
    } as const;
}

export function buildDataLikeDai(
    name: string,
    version: string,
    chainId: number,
    verifyingContract: string,
    holder: string,
    spender: string,
    nonce: string,
    allowed: boolean,
    expiry: string = defaultDeadline.toString(),
) {
    return {
        types: { Permit: DaiLikePermit },
        domain: { name, version, chainId, verifyingContract },
        message: { holder, spender, nonce, expiry, allowed },
    } as const;
}

/*
 * @param permitContract The contract object with ERC20Permit type and token address for which the permit creating.
 */
export async function getPermit(
    owner: SignerWithAddress,
    permitContract: Contract,
    tokenVersion: string,
    chainId: number,
    spender: string,
    value: string,
    deadline = defaultDeadline.toString(),
) {
    const nonce = await permitContract.nonces(owner.address);
    const name = await permitContract.name();
    const data = buildData(
        name,
        tokenVersion,
        chainId,
        permitContract.address,
        owner.address,
        spender,
        value,
        nonce.toString(),
        deadline,
    );
    const signature = await owner._signTypedData(data.domain, data.types, data.message);
    const { v, r, s } = splitSignature(signature);
    const permitCall = permitContract.interface.encodeFunctionData('permit', [
        owner.address,
        spender,
        value,
        deadline,
        v,
        r,
        s,
    ]);
    return cutSelector(permitCall);
}

/*
 * @param permitContract The contract object with ERC20PermitLikeDai type and token address for which the permit creating.
 */
export async function getPermitLikeDai(
    holder: SignerWithAddress,
    permitContract: Contract,
    tokenVersion: string,
    chainId: number,
    spender: string,
    allowed: boolean,
    expiry = defaultDeadline.toString(),
) {
    const nonce = await permitContract.nonces(holder.address);
    const name = await permitContract.name();
    const data = buildDataLikeDai(
        name,
        tokenVersion,
        chainId,
        permitContract.address,
        holder.address,
        spender,
        nonce.toString(),
        allowed,
        expiry,
    );
    const signature = await holder._signTypedData(data.domain, data.types, data.message);
    const { v, r, s } = splitSignature(signature);
    const permitCall = permitContract.interface.encodeFunctionData(
        'permit(address,address,uint256,uint256,bool,uint8,bytes32,bytes32)',
        [holder.address, spender, nonce, expiry, allowed, v, r, s],
    );
    return cutSelector(permitCall);
}

export function withTarget(target: bigint | string, data: bigint | string) {
    return target.toString() + trim0x(data);
}
