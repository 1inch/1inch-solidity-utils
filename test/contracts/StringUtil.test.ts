import { expect } from '../../src/prelude';
import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';
import { ethers } from 'hardhat';

describe('StringUtil', async () => {
    const uint256TestValue = '0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF';
    const uint128TestValue = '0x00000000000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF';
    const veryLongArray = '0xffffffffffffffafafafbcbcbcbcbdeded' + 'aa'.repeat(50);
    const extremelyLongArray = '0x' + '0f'.repeat(1000);
    const emptyBytes = '0x';
    const singleByte = '0xaf';
    const randomBytes = '0x01de89fff130adeaad';
    const sameBytesShort = '0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';
    const sameBytesLong = '0x' + 'aa'.repeat(1000);

    const deployStringUtilTest = async () => {
        const StringUtilTest = await ethers.getContractFactory('StringUtilTest');
        const stringUtilTest = await StringUtilTest.deploy();
        return { stringUtilTest };
    };

    describe('Validity', async () => {
        it('Uint 256', () => test(uint256TestValue));

        it('Uint 128', () => test(uint128TestValue));

        it('Very long byte array', () => testBytes(veryLongArray));

        it('Extremely long byte array', () => testBytes(extremelyLongArray));

        it('Empty bytes', () => testBytes(emptyBytes));

        it('Single byte', () => testBytes(singleByte));

        it('Random bytes', () => testBytes(randomBytes));

        it('Same bytes short', () => testBytes(sameBytesShort));

        it('Same bytes long', () => testBytes(sameBytesLong));

        it('Single byte naive', () => testIncorrectGas(singleByte, 2000));

        const test = async (value: string) => {
            const { stringUtilTest } = await loadFixture(deployStringUtilTest);
            const result = await stringUtilTest.toHex(value, 0);
            const naiveResult = await stringUtilTest.toHexNaive(value, 0);
            expect(result.toLowerCase()).to.be.equal(value.toLowerCase());
            expect(result.toLowerCase()).to.be.equal(naiveResult.toLowerCase());
        };

        const testBytes = async (value: string) => {
            const { stringUtilTest } = await loadFixture(deployStringUtilTest);
            const result = await stringUtilTest.toHexBytes(value, 0);
            const naiveResult = await stringUtilTest.toHexNaiveBytes(value, 0);
            expect(result.toLowerCase()).to.be.equal(value.toLowerCase());
            expect(result.toLowerCase()).to.be.equal(naiveResult.toLowerCase());
        };

        const testIncorrectGas = async (value: string, expectedGas: number) => {
            const { stringUtilTest } = await loadFixture(deployStringUtilTest);
            await expect(stringUtilTest.toHexNaiveBytes(value, expectedGas)).to.be.revertedWithCustomError(
                stringUtilTest,
                'GasCostDiffers',
            );
        };
    });

    describe('Gas usage @skip-on-coverage', async () => {
        it('Uint 256', () => testGasUint256(uint256TestValue, 907));

        it('Uint 256 naive', () => testGasNaiveUint256(uint256TestValue, 14175));

        it('Uint 256 as bytes', () => testGasBytes(uint256TestValue, 782));

        it('Uint 256 as bytes naive', () => testGasNaiveBytes(uint256TestValue, 14050));

        it('Uint 128', () => testGasUint256(uint128TestValue, 907));

        it('Uint 128 naive', () => testGasNaiveUint256(uint128TestValue, 14175));

        it('Very long byte array gas', () => testGasBytes(veryLongArray, 1964));

        it('Very long byte array gas naive', () => testGasNaiveBytes(veryLongArray, 28972));

        it('Extremely long byte array gas', () => testGasBytes(extremelyLongArray, 19121));

        it('Extremely long byte array gas naive', () => testGasNaiveBytes(extremelyLongArray, 426795));

        it('Empty bytes', () => testGasBytes(emptyBytes, 191));

        it('Empty bytes naive', () => testGasNaiveBytes(emptyBytes, 406));

        it('Single byte', () => testGasBytes(singleByte, 782));

        it('Single byte naive', () => testGasNaiveBytes(singleByte, 832));

        const testGasUint256 = async (value: string, expectedGas: number) => {
            const { stringUtilTest } = await loadFixture(deployStringUtilTest);
            await stringUtilTest.toHex(value, expectedGas);
        };

        const testGasBytes = async (value: string, expectedGas: number) => {
            const { stringUtilTest } = await loadFixture(deployStringUtilTest);
            await stringUtilTest.toHexBytes(value, expectedGas);
        };

        const testGasNaiveUint256 = async (value: string, expectedGas: number) => {
            const { stringUtilTest } = await loadFixture(deployStringUtilTest);
            await stringUtilTest.toHexNaive(value, expectedGas);
        };

        const testGasNaiveBytes = async (value: string, expectedGas: number) => {
            const { stringUtilTest } = await loadFixture(deployStringUtilTest);
            await stringUtilTest.toHexNaiveBytes(value, expectedGas);
        };
    });
});
