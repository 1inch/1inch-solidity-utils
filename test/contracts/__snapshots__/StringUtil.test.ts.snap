// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringUtil Gas usage Empty bytes 1`] = `22305n`;

exports[`StringUtil Gas usage Empty bytes naive 1`] = `22651n`;

exports[`StringUtil Gas usage Extremely long byte array gas 1`] = `60428n`;

exports[`StringUtil Gas usage Extremely long byte array gas naive 1`] = `536349n`;

exports[`StringUtil Gas usage Single byte 1`] = `22982n`;

exports[`StringUtil Gas usage Single byte naive 1`] = `23301n`;

exports[`StringUtil Gas usage Uint 128 1`] = `22815n`;

exports[`StringUtil Gas usage Uint 128 naive 1`] = `38302n`;

exports[`StringUtil Gas usage Uint 256 1`] = `23007n`;

exports[`StringUtil Gas usage Uint 256 as bytes 1`] = `23520n`;

exports[`StringUtil Gas usage Uint 256 as bytes naive 1`] = `39097n`;

exports[`StringUtil Gas usage Uint 256 naive 1`] = `38494n`;

exports[`StringUtil Gas usage Very long byte array gas 1`] = `25412n`;

exports[`StringUtil Gas usage Very long byte array gas naive 1`] = `57177n`;
